### ISSUE TITLE
onnx模型转为ncnn模型后推理结果不正确且与onnx存在明显偏差

### ISSUE THREAD

**Issue by bestpower** [Tencent/ncnn#2800]
1、通过pytorch训练出的模型转为onnx和onnx-simply模型后进行推理对比验证无误；
2、通过ncnn相关工具将onnx转为ncnn模型后部署到安卓段测试，推理结果经sofmax后出现了明显偏差：
onnx：9.9988e-01, 1.1743e-04
ncnn：0.370259, 0.629741
1）排查输入数据与ncnn预处理均没有问题
2）通过Netron工具对比onnx和ncnn两种模型各节点与输入输出未发现问题，说明模型转换应无问题
3）通过在安卓端逐个打印ncnn模型中间节点输出数据，发现在节点**BLOB_692**和**BLOB_693**有输出不全的情况，后面的节点输出也会受此影响出现问题，说明ncnn在推理模型过程中出现了问题，麻烦帮忙分析一下，模型详见附件文件。
[model.zip](https://github.com/Tencent/ncnn/files/6233444/model.zip)

**Comment by bestpower** [comment]
其中使用的ncnn代码及安卓库的commitid均为为912e81d086309cc5fe7e2de8d24c11a695f7e82a

**Comment by bestpower** [comment]
我在安卓端的jni核心推理分析代码如下：
`
ncnn::Extractor ex = MaskDetnet.create_extractor();
ex.set_num_threads(threadnum);
ex.set_light_mode(true);
ex.input(face_mask_detect_param_id::BLOB_data, img_);     // 输入节点

for(int i=1; i<208; i++)  // 遍历输出打印所有中间节点数据
{
	ncnn::Mat middleOut;
	int middle_status = ex.extract(i, middleOut);
	if(middle_status==0) {
		middleOut = middleOut.reshape(middleOut.w * middleOut.h * middleOut.c);
		unsigned int out_size = (unsigned int) (middleOut.w);
		LOGE("===== Face mask %d middle result start =====", i);
		for (int l = 0; l < out_size; l++) {
			LOGE("Face mask %d middle result = %f", i, middleOut[l]);
		}
		LOGE("===== Face mask %d middle result end =====", i);
	}else{
		LOGE("===== Face mask %d middle result fail =====", i);
	}
}

ncnn::Mat out;
int status = ex.extract(face_mask_detect_param_id::BLOB_fc, out);     // 输出节点
if (status == 0) {
	out = out.reshape(out.w * out.h * out.c);
	//概率化处理
	{
		ncnn::Layer* softmax = ncnn::create_layer("Softmax");
		ncnn::ParamDict pd;
		softmax->load_param(pd);
		softmax->forward_inplace(out, MaskDetnet.opt);
		delete softmax;
	}
	LOGE("Face mask result = %f, %f", out[0], out[1]); // 打印最终输出节点数据
}else{
	LOGE("Ncnn face mask extract failed, status = %d", status);
}
`
目前的日志分析结果为：
1、从节点1（BLOB_251）开始打印，在打印至节点序号202（BLOB_692）时出现打印不全的情况，且后续节点也未输出任何数据，最终输出节点（BLOB_fc）也未有任何输出；
2、从节点202（BLOB_692）开始打印，在打印至节点203（BLOB_693）时出现打印不全的情况，且后续节点也未输出任何数据，最终输出节点（BLOB_fc）也未有任何输出；
3、从节点203（BLOB_693）开始打印，后续节点和输出节点均可打印完全，但输出结果与onnx模型推理相差甚远。

**Comment by bestpower** [comment]
另外，我在安卓端还测试了一些其他能正确推理出结果的ncnn模型，以同样的方式打印中间节点的输出数据，只要有输出的节点，都是可以完全打印出来的，不会出现打印不全的情况，所以该问题应该是ncnn调用库中的代码存在问题，麻烦根据我现有提供的信息帮忙分析一下，如在分析过程中还需要提供其他信息，可说明。

**Comment by bestpower** [comment]
**问题复现：**
1、将训练好的pytorch模型通过工具转换为ncnn模型（pth->onnx->onnx-sim->ncnn）
2、将ncnn模型部署到安卓端应用
3、使用相同测试图片分别到pytorch、onnx和ncnn模型进行推理验证（pytorch和onnx推理结果一致，而ncnn推理结果不准确且与其他两个模型相差很大）
4、根据FAQ-ncnn-produce-wrong-result检查ncnn部署代码，未发现任何问题
5、在安卓端jni代码中添加模型相关节点日志信息打印
6、推理后发现三个问题（
      **1）、从节点1（BLOB_251）开始打印，在打印至节点序号202（BLOB_692）时出现打印不全的情况，且后续节点也未输出任何数据，最终输出节点（BLOB_fc）也未有任何输出；
      2）、从节点202（BLOB_692）开始打印，在打印至节点203（BLOB_693）时出现打印不全的情况，且后续节点也未输出任何数据，最终输出节点（BLOB_fc）也未有任何输出；
      3）、从节点203（BLOB_693）开始打印，后续节点和输出节点均可打印完全，但输出结果与onnx模型推理相差甚远。
）**
7、在其他一些能正确推理出结果的ncnn模型重复5、6步骤，未发现节点信息打印不全的情况
据此，说明可能的两点问题：
**1）onnx转ncnn过程存在问题
2、ncnn推理存在问题**

**Comment by bestpower** [comment]
@nihui 所有问题信息如以上记录，该如何解决，望回复，如有疑问可一并说明！

**Comment by nihui** [comment]
你好，模型转换上基本没有问题，但是最后的 Pooling，ncnn 输出是一维数据，不需要再 Reduction
把 param 文件最后的两行 Reduction 改为 Noop 即可获得与 onnx 一致的结果
```
HardSwish        Div_413                  1 1 694 700 0=1.666667e-01 1=5.000000e-01
Noop        ReduceMean_414           1 1 700 701 0=3 1=0 -23303=1,3 4=0
Noop        ReduceMean_415           1 1 701 702 0=3 1=0 -23303=1,2 4=0
InnerProduct     Gemm_416                 1 1 702 fc 0=2 1=1 2=20
```

**Comment by bestpower** [comment]
> 你好，模型转换上基本没有问题，但是最后的 Pooling，ncnn 输出是一维数据，不需要再 Reduction
> 把 param 文件最后的两行 Reduction 改为 Noop 即可获得与 onnx 一致的结果
> 
> ```
> HardSwish        Div_413                  1 1 694 700 0=1.666667e-01 1=5.000000e-01
> Noop        ReduceMean_414           1 1 700 701 0=3 1=0 -23303=1,3 4=0
> Noop        ReduceMean_415           1 1 701 702 0=3 1=0 -23303=1,2 4=0
> InnerProduct     Gemm_416                 1 1 702 fc 0=2 1=1 2=20
> ```
你好，按你说的这样改了下，自测了一遍应该是可以了，不过我还有一个疑问，就是按这样的改法说明模型转换在最后Pooling这两层还是有问题的吧，如果以后再遇到类似的问题，都需要这样改param文件？是否需要在onnx2ncnn方法里做相应的修正？

**Comment by nihui** [comment]
针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn
In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn

```shell
pip install pnnx
pnnx model.onnx inputshape=[1,3,224,224]
```

详细参考文档
Detailed reference documentation
https://github.com/pnnx/pnnx
https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx
